samtranslator/model/appsync.py

```diff
@@ -40,3 +73,22 @@ class GraphQLSchema(Resource):
     ApiId: Intrinsicable[str]
     Definition: Optional[str]
     DefinitionS3Location: Optional[str]
+
+
+class DataSource(Resource):
+    resource_type = "AWS::AppSync::DataSource"
+    property_types = {
+        "ApiId": PropertyType(True, IS_STR),
```

Comment:
Think we use `Property` and `PassThroughProperty` now instead of `PropertyType`. https://github.com/aws/serverless-application-model/blob/develop/DEVELOPMENT_GUIDE.md#development-guidelines

Reply:
Whoops, forgot to reflect these changes in this module like for `sam_resources.py`. good catch

Reply:
For some context, for SAM properties, we want to avoid `PropertyType` since (1) it avoids [sneaky intrinsics-related bugs](https://github.com/aws/serverless-application-model/pull/2495#discussion_r976715242) since we no longer need to resolve them (thanks `AWS::LanguageExtensions`) and (2) ensures we don't wrongly fail valid templates by supporting any values supported by the underlying property (which we don't have control over).

For CloudFormation resources (like in this case), it's arguably a little different, since we're the producer and sometimes care and know exactly what types we expected. I don't have anything to propose currently, but thinking whether these should all be `PassThroughProperty`, whether we even care about strictly typing these, or whether `PropertyType` here makes sense... 🤔

Reply:
So the problem with this whole `Resource` thing is that it's not statically typed. It's runtime type checks, which can cause runtime failures:

https://github.com/aws/serverless-application-model/blob/83306d38ff957152f7f274eef3b9b9abd873101f/samtranslator/model/types.py#L34-L40

That's why I'm a little more comfortable preferring lenient "types" such as `PropertyType` for producers which "supports" intrinsics by defaulting `self.supports_intrinsics` to true:

https://github.com/aws/serverless-application-model/blob/c39c2807bbf327255de8abed8b8150b18c60f053/samtranslator/model/__init__.py#L306-L308

It avoids risk of us prematurely failing something valid. And if there really is a problem, the CloudFormation resource knows better and will fail anyway.

So ideally the producer should be all regular, static Python types. Not sure how big of a task that is. But until then being more lenient (e.g. `PropertyType` or `PassThroughProperty`) is probably better.

Reply:
We're moving ahead with a different approach for type checking in CFN resources, since what we currently use can be  misleading. View this PR for more info:

https://github.com/aws/serverless-application-model/pull/2907